src/server/rollback.cpp

```diff
@@ -708,108 +49,116 @@ bool RollbackManager::isActorGuess()
 	return current_actor_is_guess;
 }
 
-void RollbackManager::setActor(const std::string & actor, bool is_guess)
+void RollbackManager::setActor(const std::string &actor, bool is_guess)
 {
 	current_actor = actor;
 	current_actor_is_guess = is_guess;
 }
 
-std::string RollbackManager::getSuspect(v3s16 p, float nearness_shortcut,
-		float min_nearness)
+std::string RollbackManager::getSuspect(v3s16 p, float nearness_shortcut, float min_nearness)
 {
-	if (!current_actor.empty()) {
+	if (!current_actor.empty())
 		return current_actor;
-	}
+
 	time_t cur_time = time(0);
 	time_t first_time = cur_time - (100 - min_nearness);
 	RollbackAction likely_suspect;
 	float likely_suspect_nearness = 0;
-	for (auto i = action_latest_buffer.rbegin();
-	     i != action_latest_buffer.rend(); ++i) {
-		if (i->unix_time < first_time) {
+
+	for (auto i = action_latest_buffer.rbegin(); i != action_latest_buffer.rend(); ++i) {
+		if (i->unix_time < first_time)
 			break;
-		}
-		if (i->actor.empty()) {
+		if (i->actor.empty())
 			continue;
-		}
-		// Find position of suspect or continue
+
 		v3s16 suspect_p;
-		if (!i->getPosition(&suspect_p)) {
+		if (!i->getPosition(&suspect_p))
 			continue;
-		}
-		float f = getSuspectNearness(i->actor_is_guess, suspect_p,
-					     i->unix_time, p, cur_time);
+
+		float f = getSuspectNearness(i->actor_is_guess, suspect_p, i->unix_time, p, cur_time);
 		if (f >= min_nearness && f > likely_suspect_nearness) {
 			likely_suspect_nearness = f;
 			likely_suspect = *i;
-			if (likely_suspect_nearness >= nearness_shortcut) {
+			if (likely_suspect_nearness >= nearness_shortcut)
 				break;
-			}
 		}
 	}
-	// No likely suspect was found
-	if (likely_suspect_nearness == 0) {
+
+	if (likely_suspect_nearness == 0)
 		return "";
-	}
-	// Likely suspect was found
+
 	return likely_suspect.actor;
 }
 
-
 void RollbackManager::flush()
 {
-	sqlite3_exec(db, "BEGIN", NULL, NULL, NULL);
+	if (action_todisk_buffer.empty())
+		return;
 
-	for (auto iter = action_todisk_buffer.begin();
-			iter != action_todisk_buffer.end();
-			++iter) {
-		if (iter->actor.empty()) {
-			continue;
+	beginSaveActions();
+	try {
+		for (const auto &a : action_todisk_buffer) {
+			if (a.actor.empty())
+				continue;
+			persistAction(a);
 		}
-
-		registerRow(actionRowFromRollbackAction(*iter));
+		endSaveActions();
+		action_todisk_buffer.clear();
+	} catch (...) {
+		rollbackSaveActions();
```

Comment:
this seems somewhat assuming of a database-like backend. I would have expected for `flush` to be fully implemented in the child class.